USB Device next cdc acm can get in a deadlock when the host doesn't pull the endpoint data

SynchronicIT opened the pull request:
Fixes #78104 

Current situation before the fix:
- The cdc_acm code starts claiming buffers to hand over to the lower layer driver for TX. 
- The lower layer drivers puts the data into the endpoint for the application / host
- lower layer gives a request callback when the endpoint is read
- cdc_acm layer frees the buffer. 

When the endpoint is not read, due to no terminal application running, the last two steps doesn't occur and the cdc_acm simply occupies all available buffers, followed by rescheduling the same work task for tx_fifo directly, creating an endless loop. 

By checking if a TX task is ongoing, the new work task scheduled as soon as the callback is given. 

There is no timeout used in this case, as I couldn't find an interface to configure a timeout.

SynchronicIT commented on subsys/usb/device_next/class/usbd_cdc_acm.c:
```diff
@@ -535,14 +551,20 @@ static void cdc_acm_tx_fifo_handler(struct k_work *work)
 	buf = cdc_acm_buf_alloc(cdc_acm_get_bulk_in(c_data));
 	if (buf == NULL) {
-		cdc_acm_work_submit(&data->tx_fifo_work);
+		// cdc_acm_work_submit(&data->tx_fifo_work);
```
resolved in updated pull-request

tmon-nordic commented:
Is this essentially https://github.com/zephyrproject-rtos/zephyr/issues/76642 that is fixed by https://github.com/zephyrproject-rtos/zephyr/pull/75439 or some other issue?

SynchronicIT commented:
@tmon-nordic : The fix does take in account that poll could also be used, but I haven't validated that this would fix this mixed usage.

jfischer-no commented on subsys/usb/device_next/class/usbd_cdc_acm.c:
```diff
@@ -48,6 +48,7 @@ UDC_BUF_POOL_DEFINE(cdc_acm_ep_pool,
 #define CDC_ACM_IRQ_TX_ENABLED		3
 #define CDC_ACM_RX_FIFO_BUSY		4
 #define CDC_ACM_LOCK			5
+#define CDC_ACM_TX_FIFO_BUSY		6
```
I do not want to introduce TX_FIFO_BUSY, we can handle it by using delayed work, which is needed anyway, see https://github.com/zephyrproject-rtos/zephyr/pull/75439/commits/741d8a7803c9417f755bbfa98b1f4f6fdf98d668.

jfischer-no commented on subsys/usb/device_next/class/usbd_cdc_acm.c:
```diff
@@ -217,6 +218,11 @@ static int usbd_cdc_acm_request(struct usbd_class_data *const c_data,
 
+		if (bi->ep == cdc_acm_get_bulk_in(c_data)) {
+			atomic_clear_bit(&data->state, CDC_ACM_TX_FIFO_BUSY);
+			LOG_ERR("TX ERROR !!");
```
There is no need to log an error here.

jfischer-no commented on subsys/usb/device_next/class/usbd_cdc_acm.c:
```diff
@@ -236,9 +242,19 @@ static int usbd_cdc_acm_request(struct usbd_class_data *const c_data,
+		atomic_clear_bit(&data->state, CDC_ACM_TX_FIFO_BUSY);
+		
+
+		if (ring_buf_is_empty(data->tx_fifo.rb)) {
+			/* Raise TX ready interrupt */
+			if (data->cb) {
+				cdc_acm_work_submit(&data->irq_cb_work);
+			}
+		} else {
+			/* Queue pending TX data on IN endpoint */
+			cdc_acm_work_submit(&data->tx_fifo_work);
```
No need to do anything other than trigger irq_cb_work.

SynchronicIT committed:
usb: device_next: cdc_acm: Prevent polling for buffer in TX

In case the host doesn't pull the new data from the endpoint, the work task
would schedule itself again delayed (at the max. priority). When there is
no terminal program or active application reading the endpoint this
results in a constant polling of the endpoint burning up to 5% of the
CPU cycles.

By using a atomic flag for tx busy, the polling is solved and changed into
a postponed execution of the next work task which saves up to 5% of
CPU cycles and allows a better real-time behavior for other tasks.

Secondly, if the TX interrupt is disabled but there is still data in the TX
FIFO (ring buffer), the implementation will continue to trigger subsequent
TX work and attempt to flush the data to the host.

Signed-off-by: Vincent van der Locht <vincent@synchronicit.nl>

jfischer-no reviewed (commented):
@SynchronicIT Can you please check commit message? Otherwise LGTM.

>In case the host doesn't pull the new data from the endpoint, the work task
would schedule itself again delayed (at the max. priority). When there is
no terminal program or active application. This commit changes this from
a polling behavior to

^

>By using a atomic flag for tx busy, the polling is solved and changed into
a postponed  execution of the next work taks which saves up to 5% of
CPU cycles.

jfischer-no commented on subsys/usb/device_next/class/usbd_cdc_acm.c:
```diff
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2022 Nordic Semiconductor ASA
+ * Copyright (c) 2024 SynchronicIT BV
```
I do not think that this commit is worthy of a claim here.

SynchronicIT commented on subsys/usb/device_next/class/usbd_cdc_acm.c:
```diff
@@ -548,6 +561,11 @@ static void cdc_acm_tx_fifo_handler(struct k_work *work)
+
 	buf = cdc_acm_buf_alloc(cdc_acm_get_bulk_in(c_data));
 	if (buf == NULL) {
 		cdc_acm_work_schedule(&data->tx_fifo_work, K_MSEC(1));
```
with the current implementation, having dedicated buffers for tx and rx, the buf == NULL case should never occur using the TX_BUSY flags. (both buffers are protected by a flag). 
For future changes (more buffers allocated) to go completely wrong, I added the extra clear statement. 

I also swapped the buffer free and flag free, so the flag is cleared after the buffer is cleared (and thus not busy anymore)

SynchronicIT commented:
@tmon-nordic Do you want me to add this to the commit comment? or leave it as it is. For me no problem, but it triggers potentially the review and commit checks again.

tmon-nordic commented:
> @tmon-nordic Do you want me to add this to the commit comment? or leave it as it is. For me no problem, but it triggers potentially the review and commit checks again.

I'd appreciate it if you do change the commit message. The commit message will stay there and it will be the first thing people looking through the CDC ACM class implementation history will notice. I am not sure if my review will disappear after merely commit message change, but I will re-approve as soon as I see the notification.

jfischer-no commented:
From the commit message:

>Secondly, under the current structure of usbd_cdc_acm, there is a
possibility that TX data is enqueued to UDC driver before the previous
transaction is completed, hence causing dropped data. By using a
atomic flag for tx busy, no data is dropped and will only be enqueued
when there is room to forward the data.

I am still confused by this statement, where is the data dropped, can you point me to the line. It is fine to enqueue multiple transfers!

SynchronicIT commented:
Hi @jfischer-no

I added the additional comments to the commit message on request of @tmon-nordic so that both mentioned issues are referred to. The initial goal of the commit was first of all solve the original deadlock and with the other patches, improve the performance. 
I'm hoping that we can get this merged into the main branch as soon as possible. And whatever non-technical part needs to be done, just agree together and tell me, so we can get it done. (and sorry for my direct Dutch approach...)

jfischer-no commented:
> I added the additional comments to the commit message on request of @tmon-nordic so that both mentioned issues are referred to. The initial goal of the commit was first of all solve the original deadlock and with the other patches, improve the performance. I'm hoping that we can get this merged into the main branch as soon as possible. And whatever non-technical part needs to be done, just agree together and tell me, so we can get it done. (and sorry for my direct Dutch approach...)

I have seen this [request](https://github.com/zephyrproject-rtos/zephyr/pull/78160#issuecomment-2405004382). That is addressed in https://github.com/zephyrproject-rtos/zephyr/pull/78160/commits/adfb154b253f0c4a7ff6d5698a8ee3ddeda73ab1#diff-55bd613139b01a65e6f92a663ef66b84b0041b280ad89aa53a0afd517c86b059R261-R263. I think the commit message could be:


"In case the host doesn't pull the new data from the endpoint, the work task
would schedule itself again delayed (at the max. priority). When there is
no terminal program or active application reading the endpoint this
results in a constant polling of the endpoint burning up to 5% of the
CPU cycles.

By using a atomic flag for tx busy, the polling is solved and changed into
a postponed execution of the next work task which saves up to 5% of
CPU cycles and allows a better real-time behavior for other tasks.

_Secondly, if the TX interrupt is disabled but there is still data in the TX
FIFO (ring buffer), the implementation will continue to trigger subsequent
TX work and attempt to flush the data to the host._"
